gsk4/src/render_node.rs

```diff
@@ -0,0 +1,171 @@
+                Ok(from_glib_full(self.to_glib_full()))
+            } else {
+                Err(self)
+            }
+        }
+    }
+
+    pub fn downcast_ref<T: IsARenderNode>(&self) -> Option<&T> {
+        unsafe {
+            if self.get_type() == T::static_type() {
+                Some(&*(self as *const RenderNode as *const T))
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn get_type(&self) -> glib::Type {
+        Self::static_type()
+    }
+
+    pub fn draw(&self, cr: &cairo::Context) {
+        unsafe {
+            ffi::gsk_render_node_draw(self.to_glib_none().0, mut_override(cr.to_glib_none().0));
+        }
+    }
+
+    pub fn get_bounds(&self) -> graphene::Rect {
+        unsafe {
+            let mut bounds = graphene::Rect::uninitialized();
+            ffi::gsk_render_node_get_bounds(self.to_glib_none().0, bounds.to_glib_none_mut().0);
+            bounds
+        }
+    }
+
+    pub fn get_node_type(&self) -> RenderNodeType {
+        unsafe { from_glib(ffi::gsk_render_node_get_node_type(self.to_glib_none().0)) }
+    }
+
+    pub fn serialize(&self) -> Option<glib::Bytes> {
+        unsafe { from_glib_full(ffi::gsk_render_node_serialize(self.to_glib_none().0)) }
+    }
+
+    pub fn write_to_file(&self, filename: &str) -> Result<(), glib::Error> {
+        unsafe {
+            let mut error = ptr::null_mut();
+            let _ = ffi::gsk_render_node_write_to_file(
+                self.to_glib_none().0,
+                filename.to_glib_none().0,
+                &mut error,
+            );
+            if error.is_null() {
+                Ok(())
+            } else {
+                Err(from_glib_full(error))
+            }
+        }
+    }
+}
+
+pub trait IsARenderNode:
+    StaticType
+    + FromGlibPtrFull<*mut ffi::GskRenderNode>
+    + std::convert::AsRef<crate::RenderNode>
+    + 'static
+{
+    fn upcast(self) -> RenderNode;
+    fn upcast_ref(&self) -> &RenderNode;
+}
+
+pub const NONE_RENDER_NODE: Option<&RenderNode> = None;
+
+impl fmt::Display for RenderNode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("RenderNode")
+    }
+}
+
+#[macro_export]
+macro_rules! define_render_node {
+    ($rust_type:ident, $display_name:expr, $ffi_type:path, $get_type:path) => {
+        // Can't use get_type here as this is not a boxed type but another fundamental type
+        glib::wrapper! {
+            #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+            pub struct $rust_type(Shared<$ffi_type>);
+
+            match fn {
+                ref => |ptr| ffi::gsk_render_node_ref(ptr as *mut ffi::GskRenderNode) as *mut $ffi_type,
+                unref => |ptr| ffi::gsk_render_node_unref(ptr as *mut ffi::GskRenderNode),
+                get_type => || $get_type(),
+            }
+        }
+
+        unsafe impl Send for $rust_type {}
+        unsafe impl Sync for $rust_type {}
```

Comment:
Are the render nodes actually Send?

Reply:
I guess I'm not sure.  I removed it so it should be the same as it was before my change.